src/Http/Http.Extensions/src/DefaultProblemDetailsWriter.cs

```diff
@@ -43,20 +45,20 @@ public bool CanWrite(ProblemDetailsContext context)
     }
 
     [UnconditionalSuppressMessage("Trimming", "IL2026",
-        Justification = "JSON serialization of ProblemDetails.Extensions might require types that cannot be statically analyzed and we need to fallback" +
-        "to reflection-based. The ProblemDetailsConverter is marked as RequiresUnreferencedCode already.")]
+        Justification = "JSON serialization of ProblemDetails.Extensions might require types that cannot be statically analyzed. The property is annotated with a warning")]
     [UnconditionalSuppressMessage("Trimming", "IL3050",
-        Justification = "JSON serialization of ProblemDetails.Extensions might require types that cannot be statically analyzed and we need to fallback" +
-        "to reflection-based. The ProblemDetailsConverter is marked as RequiresDynamicCode already.")]
+        Justification = "JSON serialization of ProblemDetails.Extensions might require types that cannot be statically analyzed. The property is annotated with a warning")]
     public ValueTask WriteAsync(ProblemDetailsContext context)
     {
         var httpContext = context.HttpContext;
         ProblemDetailsDefaults.Apply(context.ProblemDetails, httpContext.Response.StatusCode);
         _options.CustomizeProblemDetails?.Invoke(context);
 
-        if (context.ProblemDetails.Extensions is { Count: 0 })
+        // Use source generation serialization in two scenarios:
+        // 1. There are no extensions. Source generation is faster and works well with trimming.
+        // 2. Native AOT. In this case only the data types specified on ProblemDetailsJsonContext will work.
+        if (context.ProblemDetails.Extensions is { Count: 0 } || !RuntimeFeature.IsDynamicCodeSupported)
```

Comment:
@eerhardt @jkotas @vitek-karas @davidfowl Like https://github.com/dotnet/aspnetcore/pull/45604#discussion_r1050139326, this is a place where https://github.com/dotnet/runtime/issues/39806 would be useful.

Developers being able to run the "AOT path" while writing and debugging their app would help them find incompatible extension values during development instead of when they publish as AOT for production.

Reply:
Would it fail to serialize if we did that in "build"? Or would the reflection based serializer kick in?

Reply:
It would fail at runtime. This is what we're going to be doing manually today for AOT based JSON in ASP.NET Core. The reflection fallback would be disabled and our calls to resolve the JsonTypeInfo would return null.

Reply:
In that case - let's do https://github.com/dotnet/runtime/issues/39806.
We just need to "hide" it enough so that people don't use it as the first option to solve AOT problems (still better than current behavior, but ideally we want them to really fix things)